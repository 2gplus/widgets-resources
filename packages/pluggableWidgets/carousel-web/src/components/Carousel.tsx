import { createElement, ReactNode, ReactElement } from "react";
import { ObjectItem } from "mendix";
import Swiper, { SwiperOptions, Navigation, Pagination, EffectFade, Autoplay } from "swiper";
import { Swiper as SwiperReact, SwiperSlide } from "swiper/react";

export interface CarouselProps<T extends ObjectItem> {
    pagination: boolean;
    loop: boolean;
    tabIndex?: number | undefined;
    animation?: boolean;
    autoplay?: boolean;
    delay?: number;
    navigation: boolean;
    onClick: () => void;
    items: T[];
    itemRenderer: (renderWrapper: (children: ReactNode) => ReactElement, item: T) => ReactNode;
}

export function Carousel<T extends ObjectItem>(props: CarouselProps<T>): ReactElement {
    const { items, pagination, loop, tabIndex, animation, autoplay, delay, navigation, itemRenderer, onClick } = props;
    Swiper.use([Navigation, Pagination, EffectFade, Autoplay]);
    const options: SwiperOptions = {
        slidesPerView: 1,
        centeredSlides: true,
        loop,
        navigation,
        autoplay: autoplay ? { delay, stopOnLastSlide: true } : false,
        pagination: pagination ? { type: "bullets", clickable: true } : false,
        ...(animation && {
            effect: "fade",
            fadeEffect: { crossFade: true }
        })
    };
    return (
        <div className={"widget-carousel"} tabIndex={tabIndex}>
            <SwiperReact {...options} onClick={onClick}>
                {items.length > 0 &&
                    itemRenderer &&
                    items?.map((item: T, key: number) =>
                        itemRenderer((children: ReactNode) => <SwiperSlide key={key}>{children}</SwiperSlide>, item)
                    )}
            </SwiperReact>
        </div>
    );
}
