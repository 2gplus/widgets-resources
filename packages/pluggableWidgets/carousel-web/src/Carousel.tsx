import { createElement, ReactNode, useEffect, useState, useCallback, ReactElement } from "react";
import { executeAction } from "@mendix/piw-utils-internal";
import { CarouselContainerProps } from "../typings/CarouselProps";
import { Carousel as CarouselComponent } from "./components/Carousel";
import loadingCircleSvg from "./ui/loading-circle.svg";
import "./ui/Carousel.scss";

export function Carousel(props: CarouselContainerProps): ReactNode {
    const { dataSource, content, showPagination, loop, tabIndex, navigation, animation, delay, autoplay } = props;
    const [loading, setLoading] = useState(true);
    const onClick = useCallback(() => executeAction(props.onClickAction), [props.onClickAction]);
    useEffect(() => {
        if (dataSource?.status === "available") {
            setLoading(false);
        }
    }, [dataSource]);

    const itemRenderer = useCallback((renderWrapper, item) => renderWrapper(content?.get(item)), [content]);
    const renderCarousel = (): ReactElement => {
        return (
            <CarouselComponent
                pagination={showPagination}
                loop={loop}
                tabIndex={tabIndex}
                animation={animation}
                autoplay={autoplay}
                delay={delay}
                navigation={navigation}
                itemRenderer={itemRenderer}
                items={dataSource?.items ?? []}
                onClick={onClick}
            />
        );
    };
    const renderLoading = (): ReactNode => {
        return <img src={loadingCircleSvg} className="widget-carousel-loading-spinner" alt="" aria-hidden />;
    };
    return loading ? renderLoading() : renderCarousel();
}
