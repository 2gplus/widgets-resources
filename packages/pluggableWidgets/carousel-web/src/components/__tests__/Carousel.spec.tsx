import { createElement } from "react";
import { Carousel, CarouselProps } from "../Carousel";
import { ObjectItem } from "mendix";
import renderer from "react-test-renderer";

const itemWrapperFunction = (): CarouselProps<ObjectItem>["itemRenderer"] => (wrapper, item) => wrapper(item.id);

describe("Carousel", () => {
    const defaultCarouselProps: CarouselProps<ObjectItem> = {
        items: [],
        pagination: true,
        animation: true,
        autoplay: true,
        delay: 3000,
        loop: true,
        navigation: true,
        tabIndex: undefined,
        onClick: () => jest.fn(),
        itemRenderer: itemWrapperFunction
    };

    it("renders correctly", () => {
        const carousel = renderer.create(<Carousel {...defaultCarouselProps} />);

        expect(carousel).toMatchSnapshot();
    });
    it("renders correctly without pagination", () => {
        const carousel = renderer.create(<Carousel {...defaultCarouselProps} pagination={false} />);

        expect(carousel).toMatchSnapshot();
    });
    it("renders correctly without navigation", () => {
        const carousel = renderer.create(<Carousel {...defaultCarouselProps} navigation={false} />);

        expect(carousel).toMatchSnapshot();
    });
});
