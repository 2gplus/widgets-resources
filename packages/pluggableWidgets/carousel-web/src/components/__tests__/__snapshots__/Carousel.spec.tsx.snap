// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel renders correctly 1`] = `
<div
  className="widget-carousel"
>
  <div
    className="swiper-container"
  >
    <div
      className="swiper-button-prev"
    />
    <div
      className="swiper-button-next"
    />
    <div
      className="swiper-pagination"
    />
    <div
      className="swiper-wrapper"
    />
  </div>
</div>
`;

exports[`Carousel renders correctly without navigation 1`] = `
<div
  className="widget-carousel"
>
  <div
    className="swiper-container"
  >
    <div
      className="swiper-pagination"
    />
    <div
      className="swiper-wrapper"
    />
  </div>
</div>
`;

exports[`Carousel renders correctly without pagination 1`] = `
<div
  className="widget-carousel"
>
  <div
    className="swiper-container"
  >
    <div
      className="swiper-button-prev"
    />
    <div
      className="swiper-button-next"
    />
    <div
      className="swiper-wrapper"
    />
  </div>
</div>
`;
